### ISSUE TITLE
How to restructure a tensor according to one dimension.

### ISSUE THREAD

**Issue by XuSShuai** [tensorflow/tensorflow#24070]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Centos 7
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:no
- **TensorFlow installed from (source or binary)**:binary
- **TensorFlow version (use command below)**:1.7.0
- **Python version**:3.5.5
- **Bazel version (if compiling from source)**:no
- **GCC/Compiler version (if compiling from source)**:4.9.3
- **CUDA/cuDNN version**:no
- **GPU model and memory**:2 same GTX Titan X (Pascal), 12GB
- **Exact command to reproduce**:

### Describe the problem

How to restructure a tensor according to one dimension specified by user?

The paper "An End-to-End Deep Learning Architecture for Graph Classification" published in AAAI-18([PDF](http://www.cse.wustl.edu/~muhan/papers/AAAI_2018_DGCNN.pdf)) proposed a new novel neural network architecture for `graph` classification and a new layer called "SortPooling" as a key block of the network. The author of this paper implement the network by `torch`, i want to reproduce by tensorflow. One of the key operation is to restructure a tensor by the descend value according to some one dimension.

e.g.

assume the original tensor `a` is 
```python
[[93, 69, 38, 46]
[54, 5, 1, 13]
[96, 45, 52, 50]]
```
and we want to restructure the tensor above according to the `a[:, 3]`, so output:
```python
[54, 5, 1, 13]
[[93, 69, 38, 46]
[96, 45, 52, 50]]
```
Then remove the last row(if two line already enough), so the final result i want is:
```python
[[54, 5, 1, 13]
[93, 69, 38, 46]]
```

How to achieve this by tensorflow? Any funciton exist? the `tf.nn.top_k` seem not enough for this.

### Source code / logs
no

**Comment by FredVaugeois** [comment]
Can you explain the reason why you want to do that operator?

**Comment by XuSShuai** [comment]
Maybe i can give you a brief explanation try my best. To do the graph(not image) classification, a more specialized architecture need to be designed, so `Graph Convolution layer` and `SortPooling Layer` are proposed to form GNN network. In my opinion, what the former do is extract features of each node from his neighbors by propagation, and as the author said, "our graph convolution layer is a theoretically closer approximation to the Weisfeiler Lehman algorthm", the WL algorithm is a graph labeling algorithm, what the latter do is to sort vertex by features(the value can be view as the importance of the node) instead of summing them up(the traditional method use), and feed the result into classic convolution layer.
The `Graph Convolution layer` is easy to code base on tensorflow, it just consist of some matrix operation(i think). but the `SortPooling layer` require sort and remove some dimension.  Maybe you have to dive into the paper if you want more detail and explanation.

**Comment by FredVaugeois** [comment]
Can't you use a softmax layer or some kind of classification layer at the end to sort the tensors (which in this case, would be nodes) by importance? You could actually make a 2-network program: the first one gets the importance of the nodes, which you classify using some kind of softmax layer. Then, you can remove the layers you don't want and feed the result in another NN as you see fit. I don't know if that is applicable to your problem AT ALL, but that's my best answer given my actual knowledge of your problem...

**Comment by XuSShuai** [comment]
This requirement can be meet by `tf.gather()` combine `tf.nn.top_k()`.

```python
x = tf.constant(value=[[93, 69, 38, 46],[54, 5, 1, 13],[96, 45, 52, 50]])
sort_op = tf.gather(x, tf.nn.top_k(-x[:, 3], 3).indices)[:2]
with tf.Session() as sess:
    print(sess.run(sort_op))
```
